project/part3/src/bin/main.rs

```diff
@@ -0,0 +1,514 @@
+        mk_static!(StackResources<6>, StackResources::<6>::new()),
+        seed,
+    );
+
+    // Init network stack for STA (client connection)
+    let (sta_stack, sta_runner) = embassy_net::new(
+        sta_device,
+        sta_config,
+        mk_static!(StackResources<3>, StackResources::<3>::new()),
+        seed,
+    );
+
+    spawner.spawn(connection(controller)).ok();
+    spawner.spawn(net_task(ap_runner)).ok();
+    spawner.spawn(sta_net_task(sta_runner)).ok();
+    spawner.spawn(run_dhcp(ap_stack, gw_ip_addr)).ok();
+    spawner.spawn(run_captive_portal(ap_stack, gw_ip_addr)).ok();
+    spawner.spawn(http_client_task(sta_stack)).ok();
+
+    loop {
+        if ap_stack.is_link_up() {
+            break;
+        }
+        Timer::after(EmbassyDuration::from_millis(500)).await;
+    }
+    println!("WiFi Provisioning Portal Ready");
+    println!("1. Connect to the AP: `esp-radio`");
+    println!("2. Navigate to: http://{gw_ip_addr_str}/");
+    while !ap_stack.is_config_up() {
+        Timer::after(EmbassyDuration::from_millis(100)).await
+    }
+    ap_stack
+        .config_v4()
+        .inspect(|c| println!("ipv4 config: {c:?}"));
+
+    spawner.spawn(run_http_server(ap_stack)).ok();
+
+    // Keep main task alive
+    loop {
+        Timer::after(EmbassyDuration::from_secs(60)).await;
+    }
+}
+
+// Define the form structure for WiFi credentials
+#[derive(serde::Deserialize)]
+struct WifiForm {
+    ssid: heapless::String<32>,
+    password: heapless::String<64>,
+}
+
+// Create router with picoserve
+fn make_app() -> picoserve::Router<impl picoserve::routing::PathRouter<(), picoserve::routing::NoPathParameters>, (), picoserve::routing::NoPathParameters> {
+    picoserve::Router::new()
+        .route("/", picoserve::routing::get(home_handler))
+        .route("/save", picoserve::routing::post(save_handler))
+        .route("/generate_204", picoserve::routing::get(captive_redirect))
+        .route("/gen_204", picoserve::routing::get(captive_redirect))
+        .route("/ncsi.txt", picoserve::routing::get(captive_redirect))
+        .route("/connecttest.txt", picoserve::routing::get(captive_redirect))
+}
+
+async fn home_handler() -> (picoserve::response::StatusCode, &'static [(&'static str, &'static str)], &'static str) {
+    (
+        picoserve::response::StatusCode::OK,
+        &[("Content-Type", "text/html; charset=utf-8")],
+        HOME_HTML,
+    )
+}
+
+async fn save_handler(
+    form: picoserve::extract::Form<WifiForm>
+) -> (picoserve::response::StatusCode, &'static [(&'static str, &'static str)], &'static str) {
+    println!("WiFi Credentials Received: SSID: {} | Password: {}", form.0.ssid, form.0.password);
+
+    // Send credentials to the connection task
+    let credentials = WifiCredentials {
+        ssid: form.0.ssid,
+        password: form.0.password,
+    };
+    println!("Sending credentials to connection task...");
+    WIFI_CREDENTIALS_CHANNEL.sender().send(credentials).await;
```

Comment:
Not a fan of this global state, can't we pass the channel as some context parameter through picoserve?